src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -1053,16 +1055,71 @@ public void Connect(IPAddress[] addresses, int port)
                 throw new NotSupportedException(SR.net_invalidversion);
             }
 
+// Disable CS0162: Unreachable code detected
+//
+// SuportsMultipleConnectAttempts is a constant; when false, the following lines will trigger CS0162.
+#pragma warning disable 162
             Exception lastex = null;
-            foreach (IPAddress address in addresses)
+            if (SocketPal.SupportsMultipleConnectAttempts)
```

Comment:
> All a new platform has to do for this code is declare is whether or not it supports multiple connection attempts on a socket--moving it behind the PAL would likely require a fair amount of code copying. I'd be more inclined to guard this code with #ifdefs than to move it behind the PAL.

I don't agree with the code copying part: at this time only one platform has this issue. Future extensions can be easily combined at an intermediate PAL layer: Windows (with 2 platforms), Linux will share the current code. OSX (and probably FreeBSD) will share the other classes.